### PR TITLE
Fix #3532 #4152 - Make heating/cooling coils optional for ZoneHVACLowTempVariableFlow and add missing fields to coils

### LINKED ISSUE #3532: ZoneHVACLowTempRadiantVarFlow shouldn't require a cooling coil
Currently, the constructor for zone low temp radiant systems is: ZoneHVACLowTempRadiantVarFlow (const Model &model, Schedule &availabilitySchedule, HVACComponent &heatingCoil, HVACComponent &coolingCoil)

This gets populated with Coil:Heating:LowTemperatureRadiant:VariableFlow and Coil:Cooling:LowTemperatureRadiant:VariableFlow objects.

The underlying EnergyPlus object doesn't require coils, but instead has inlet and outlet nodes for hot and chilled water as fields in the object.  OpenStudio is using these dummy coils to make it easy to connect to plant loops, inheriting from the HVACComponent object.  The issue is that cooling inlet and outlet nodes are optional in the original EnergyPlus object for a heating only system.  There isn't a way in the API to remove or ignore the cooling coil to make the translated EnergyPlus objects heating only.

Suggested solution is to make the Coil:Cooling:LowTemperatureRadiant:VariableFlow object in the ZoneHVACLowTempRadiantVarFlow constructor optional.

### LINKED ISSUE #4152: Add missing fields for CoilHeatingLowTempRadiantVarFlow and CoilCoolingLowTempRadiantVarFlow
Enhancement Request
--------------
 
This was mentioned as a comment on https://github.com/NREL/OpenStudio/issues/3532#issuecomment-492301153. I'm splitting it because we might want to do one without the other.

## Detailed Description

https://github.com/NREL/OpenStudio/issues/3532#issuecomment-492301153 says:

Additional request: add heating and cooling design fields to the OpenStudio API.

- Heating Design Capacity Method
- Heating Design Capacity
- Heating Design Capacity Per Floor Area
- Fraction of Autosized Heating Design Capacity

+ the same for cooling

## Possible Implementation

Add to IDD, add getter setters (nothing optional needed), then forward translate these.

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

- Fixes #3532
    * Make heating and cooling coils optional for ZoneHVACLowTempVariableFlow 
    * This is an **API Breaking Change**

- Fixes #4152
    * Add missing fields for CoilHeating/CoolingLowTempRadiantVarFlow. Not breaking API.

### Pull Request Author

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.

 - [x] Model API Changes / Additions
 - [x] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [x] Model API methods are tested (in `src/model/test`)
 - [x] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [x] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Checked behavior in OpenStudioApplication, adjusted policies as needed (`src/openstudio_lib/library/OpenStudioPolicy.xml`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -15,7 +15,7 @@ if(CCACHE_PROGRAM)
 endif()
 
 # Project macro can only take digits in the version, it splits these out to specific variables
-project(OpenStudio VERSION 3.1.0)
+project(OpenStudio VERSION 3.1.1)
```
Comment (jmarrec) [inline CMakeLists.txt:18]: Had to bump for VersionTranslation.

File: resources/model/OpenStudio.idd [hunk 2]
```diff
@@ -16551,42 +16551,77 @@ OS:Coil:Cooling:LowTemperatureRadiant:ConstantFlow,
...
         \reference CoilCoolingLowTempRadiantVariableFlow
         \reference ConnectionObject
-   N1 , \field Maximum Cold Water Flow
+   A3 , \field Cooling Design Capacity Method
+        \type choice
+        \key None
+        \key CoolingDesignCapacity
+        \key CapacityPerFloorArea
+        \key FractionOfAutosizedCoolingCapacity
+        \required-field
```
Comment (jmarrec) [inline resources/model/OpenStudio.idd:16568]: All fields I added made `\required-field`

File: resources/model/OpenStudio.idd [hunk 3]
```diff
@@ -40,7 +40,7 @@ OS:Version,
        \required-field
   A2, \field Version Identifier
        \type alpha
-       \default 3.1.0
+       \default 3.1.1
```
Comment (jmarrec) [inline resources/model/OpenStudio.idd:43]: Had to bump for VersionTranslation.

File: src/energyplus/ForwardTranslator/ForwardTranslateZoneHVACLowTempRadiantVarFlow.cpp [hunk 4]
```diff
@@ -84,6 +84,15 @@ namespace energyplus {
       return boost::none;
     }
 
+    boost::optional<HVACComponent> heatingCoil = modelObject.heatingCoil();
+    boost::optional<HVACComponent> coolingCoil = modelObject.coolingCoil();
+    if (!(heatingCoil.has_value() || coolingCoil.has_value())) {
+      LOG(Info,
+          modelObject.briefDescription() << " does not have either a Heating nor a Cooling Coil, it will not be translated");
+      return boost::none;
+    }
```
Comment (jmarrec) [inline src/energyplus/ForwardTranslator/ForwardTranslateZoneHVACLowTempRadiantVarFlow.cpp:93]: Don't VT if the ZoneHVACLowTemp object doesn't have at least one heating or cooling coil.

File: src/model/ZoneHVACLowTempRadiantVarFlow.hpp [hunk 5]
```diff
@@ -73,9 +76,9 @@ namespace model {
 
     Schedule availabilitySchedule() const;
 
-    HVACComponent heatingCoil() const;
+    boost::optional<HVACComponent> heatingCoil() const;
 
-    HVACComponent coolingCoil() const;
+    boost::optional<HVACComponent> coolingCoil() const;
```
Comment (jmarrec) [inline src/model/ZoneHVACLowTempRadiantVarFlow.hpp:81]: API breaking change is here.

File: src/model/ZoneHVACLowTempRadiantVarFlow.hpp [hunk 6]
```diff
@@ -58,6 +58,9 @@ namespace model {
     explicit ZoneHVACLowTempRadiantVarFlow(const Model& model, Schedule& availabilitySchedule, HVACComponent& heatingCoil,
                                            HVACComponent& coolingCoil);
 
+    // This constructor defaults the availabilitySchedule to alwaysOnDiscrete, and **does not set any heating nor cooling coil**
+    explicit ZoneHVACLowTempRadiantVarFlow(const Model& model);
```
Comment (jmarrec) [inline src/model/ZoneHVACLowTempRadiantVarFlow.hpp:62]: New **added** minimal Ctor

File: src/model/ZoneHVACLowTempRadiantVarFlow.hpp (hunk 5 above)
Comment (jmarrec) [reply src/model/ZoneHVACLowTempRadiantVarFlow.hpp:81]: @kbenne

**Comment by tijcolem** [comment]
Looks like the new cppcheck version released today (2.3-25-20201207-23616-ga9f4a14c8 2020-12-07) is broken so that is why it is failing (not due to our code). snap doesn't seem to support installing specific versions and the apt cppcheck versions is  old (1.82) and doesn't support std=c++17.  I'll look into alternatives...

**Comment by tijcolem** [comment]
as for cppcheck failing, see issue #4154

**Pull request merged**